src/services/pathfinding.service.ts

```diff
@@ -1,54 +1,164 @@
- * 4. Make a copy of the floor's travel node list and update it with the new relation found in 2. and 3.
- *
- * 5. Now that the tree is set properly. Do A* (pref. someone who did COMP472 would do this):
- *
- * 1. Initialize the open list with the initial node so (top node)
- * 2. Initialize the closed list to empty
- * 3. Repeat
- *  a) If the open list is empty, then exit with failure.
- *  b) Else, take the first node s from the open list.
- *  c) If s is a goal state, exit with success. Extract the solution path from s to so
- *  d) Else, insert s in the closed list (s has been visited /expanded)
- *  e) Insert the successors of s in the open list in a certain order if
- *  they are not already in the closed and/or open lists (to avoid
- *  cycles)
+ * Find the shortest path between two locations on a given floor.
+ * Returns an array of lines if the path is found.
+ * Returns null if no path is found.
+ * @param floorId
+ * @param start
+ * @param end
  */
 export const findPathOnFloor = (
   floorId: number,
   start: Location,
   end: Location
 ): Line[] => {
-  return [];
+  const { travelNodes } = buildingFloors.find(({ id }) => id === floorId);
+  const nodes = _.cloneDeep(travelNodes);
+
+  const edges: Line[] = traverseNodes(nodes);
+  const startEdge: Line = findClosestLine(edges, start);
+  const endEdge: Line = findClosestLine(edges, end);
+
+  const initial: TravelNode = {
+    id: nodes.length,
+    location: start,
+    children: [startEdge[0].id, startEdge[1].id]
+  };
+  const goal: TravelNode = {
+    id: nodes.length + 1,
+    location: end,
+    children: [endEdge[0].id, endEdge[1].id]
+  };
+
+  if (_.isEqual(startEdge, endEdge)) {
+    initial.children.push(goal.id);
+    goal.children.push(initial.id);
+  }
+  nodes.push(initial, goal);
+
+  nodes[startEdge[0].id].children.push(initial.id);
+  nodes[startEdge[0].id].children = nodes[startEdge[0].id].children.filter(
+    c => startEdge[1].id !== c
+  );
+
+  nodes[startEdge[1].id].children.push(initial.id);
+  nodes[startEdge[1].id].children = nodes[startEdge[1].id].children.filter(
+    c => startEdge[0].id !== c
+  );
+
+  nodes[endEdge[0].id].children.push(goal.id);
+  nodes[endEdge[0].id].children = nodes[endEdge[0].id].children.filter(
+    c => endEdge[1].id !== c
+  );
+
+  nodes[endEdge[1].id].children.push(goal.id);
+  nodes[endEdge[1].id].children = nodes[endEdge[1].id].children.filter(
+    c => endEdge[0].id !== c
+  );
+
+  const searchPath = search(nodes, initial, goal);
+
+  if (searchPath === null) return null;
+
+  const shortestPath: Line[] = [];
+  let { id } = goal;
+  while (id !== initial.id) {
+    const parentId: number = searchPath[id];
+    shortestPath.push([nodes[id], nodes[parentId]]);
+    id = parentId;
+  }
+  return shortestPath;
+};
+
+/**
+ * A* implementation for indoor pathfinding
+ * @param nodes
+ * @param initial
+ * @param goal
+ */
+const search = (
+  nodes: TravelNode[],
+  initial: TravelNode,
+  goal: TravelNode
+): {} => {
```

Comment:
This function takes in travel nodes and returns any type of object? Is it possible to be more specific with the return type here?

Reply:
We tried, and the alternative of {} that we could think of is defining a type that will only be used once, which is not the best. Please suggest more alternatives.

Reply:
Yes, it would be possible to define a type for the output